// ui-router-react/components/UISrefActive.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var react_1 = require('react');
var classNames = require('classnames');
var index_1 = require('../index');
var UISrefActive = (function (_super) {
    __extends(UISrefActive, _super);
    function UISrefActive() {
        var _this = this;
        _super.apply(this, arguments);
        // keep track of states to watch and their activeClasses
        this.states = [];
        this.activeClasses = {};
        this.addStateInfo = function (stateName, stateParams) {
            var activeClass = _this.props.class;
            var deregister = _this.addState(stateName, stateParams, activeClass);
            _this.forceUpdate();
            return deregister;
        };
        this.addState = function (stateName, stateParams, activeClass) {
            var stateService = index_1.default.instance.stateService;
            var parent = _this.context['parentUIViewAddress'];
            var stateContext = parent && parent.context || index_1.default.instance.stateRegistry.root();
            var state = stateService.get(stateName, stateContext);
            var stateHash = _this.createStateHash(stateName, stateParams);
            var stateInfo = {
                state: state || { name: stateName },
                params: stateParams,
                hash: stateHash
            };
            _this.states.push(stateInfo);
            _this.activeClasses[stateHash] = activeClass;
            return function () {
                var idx = _this.states.indexOf(stateInfo);
                if (idx !== -1)
                    _this.states.splice(idx, 1);
            };
        };
        this.createStateHash = function (state, params) {
            if (typeof state !== 'string') {
                throw new Error('state should be a string');
            }
            return params && typeof params === 'object'
                ? state + JSON.stringify(params)
                : state;
        };
        this.getActiveClasses = function () {
            var activeClasses = [];
            var stateService = index_1.default.instance.stateService;
            var exact = _this.props.exact;
            _this.states.forEach(function (s) {
                var state = s.state, params = s.params, hash = s.hash;
                if (!exact && stateService.includes(state.name, params))
                    activeClasses.push(_this.activeClasses[hash]);
                if (exact && stateService.is(state.name, params))
                    activeClasses.push(_this.activeClasses[hash]);
            });
            return activeClasses;
        };
    }
    UISrefActive.prototype.getChildContext = function () {
        return {
            parentUiSrefActiveAddStateInfo: this.addStateInfo
        };
    };
    UISrefActive.prototype.componentWillMount = function () {
        var _this = this;
        // register callback for state change
        this.deregister = index_1.default.instance.transitionService.onSuccess({}, function () { return _this.forceUpdate(); });
    };
    UISrefActive.prototype.componentWillUnmount = function () {
        this.deregister();
    };
    UISrefActive.prototype.render = function () {
        var activeClasses = this.getActiveClasses();
        return activeClasses.length > 0
            ? react_1.cloneElement(this.props.children, Object.assign({}, this.props.children.props, {
                className: classNames(this.props.children.props.className, activeClasses)
            }))
            : this.props.children;
    };
    UISrefActive.propTypes = {
        class: react_1.PropTypes.string.isRequired,
        children: react_1.PropTypes.element.isRequired
    };
    UISrefActive.contextTypes = {
        parentUIViewAddress: react_1.PropTypes.object
    };
    UISrefActive.childContextTypes = {
        parentUiSrefActiveAddStateInfo: react_1.PropTypes.func
    };
    return UISrefActive;
}(react_1.Component));
exports.UISrefActive = UISrefActive;
