// ui-router-react/index.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var ui_router_core_1 = require('ui-router-core');
exports.trace = ui_router_core_1.trace;
require('./justjs');
var justjs_1 = require('./justjs');
var ui_router_react_1 = require("./ui-router-react");
var UIView_1 = require("./components/UIView");
exports.UIView = UIView_1.UIView;
var UISref_1 = require("./components/UISref");
exports.UISref = UISref_1.UISref;
var UISrefActive_1 = require("./components/UISrefActive");
exports.UISrefActive = UISrefActive_1.UISrefActive;
// Set up view config factory
var viewConfigFactory = function (node, config) {
    return new ui_router_react_1.ReactViewConfig(node, config);
};
var UIRouterReact = (function (_super) {
    __extends(UIRouterReact, _super);
    function UIRouterReact() {
        _super.call(this);
        this.viewService.viewConfigFactory('react', viewConfigFactory);
        this.stateRegistry.decorator("views", ui_router_react_1.reactViewsBuilder);
        UIRouterReact.instance = this;
    }
    UIRouterReact.prototype.start = function () {
        this.stateRegistry.stateQueue.autoFlush(this.stateService);
        this.urlRouter.listen();
        this.urlRouter.sync();
    };
    UIRouterReact.prototype.html5Mode = function (mode) {
        if (mode === true) {
            var loc = ui_router_core_1.services.location;
            var locCfg = ui_router_core_1.services.locationConfig;
            locCfg.html5Mode = function () { return true; };
            Object.assign(loc, justjs_1.pushStateLocation);
        }
    };
    return UIRouterReact;
}(ui_router_core_1.UIRouter));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = UIRouterReact;
