// ui-router-react/components/UIView.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var react_1 = require('react');
var ui_router_core_1 = require("ui-router-core");
var index_1 = require("../index");
var id = 0;
var UIView = (function (_super) {
    __extends(UIView, _super);
    function UIView() {
        _super.apply(this, arguments);
        this.state = {
            loaded: false,
            component: 'div',
            props: {}
        };
    }
    UIView.prototype.render = function () {
        var _this = this;
        var children = this.props.children;
        var _a = this.state, component = _a.component, props = _a.props, loaded = _a.loaded;
        // register reference of child component
        // register new hook right after component has been rendered
        var stateName = this.uiViewAddress && this.uiViewAddress.context && this.uiViewAddress.context.name;
        props.ref = function (c) {
            _this.componentInstance = c;
            _this.registerUiCanExitHook(stateName);
        };
        var child = !loaded && react_1.isValidElement(children)
            ? react_1.cloneElement(children, props)
            : react_1.createElement(component, props);
        return child;
    };
    UIView.prototype.getChildContext = function () {
        return {
            parentUIViewAddress: this.uiViewAddress
        };
    };
    UIView.prototype.componentWillMount = function () {
        var router = index_1.default.instance;
        // Check the context for the parent UIView's fqn and State
        var parent = this.context['parentUIViewAddress'];
        // Not found in context, this is a root UIView
        parent = parent || { fqn: "", context: router.stateRegistry.root() };
        var name = this.props.name || "$default";
        this.uiViewData = {
            $type: 'react',
            id: ++id,
            name: name,
            fqn: parent.fqn ? parent.fqn + "." + name : name,
            creationContext: parent.context,
            configUpdated: this.viewConfigUpdated.bind(this),
            config: undefined
        };
        this.uiViewAddress = { fqn: this.uiViewData.fqn, context: undefined };
        this.deregister = router.viewService.registerUIView(this.uiViewData);
        this.setState({ id: this.uiViewData.id });
    };
    UIView.prototype.componentWillUnmount = function () {
        this.deregister();
    };
    UIView.prototype.viewConfigUpdated = function (newConfig) {
        var newComponent = newConfig && newConfig.viewDecl && newConfig.viewDecl.component;
        var trans = undefined, resolves = {};
        if (newConfig) {
            var context = newConfig.viewDecl && newConfig.viewDecl.$context;
            this.uiViewAddress = { fqn: this.uiViewAddress.fqn, context: context };
            var ctx = new ui_router_core_1.ResolveContext(newConfig.path);
            trans = ctx.getResolvable(ui_router_core_1.Transition).data;
            var stringTokens = trans.getResolveTokens().filter(function (x) { return typeof x === 'string'; });
            resolves = stringTokens.map(function (token) { return [token, trans.getResolveValue(token)]; }).reduce(ui_router_core_1.applyPairs, {});
        }
        this.uiViewData.config = newConfig;
        var props = { resolves: resolves, transition: trans };
        // attach any style or className to the rendered component
        // specified on the UIView itself
        var styleProps = {};
        if (this.props.className)
            styleProps.className = this.props.className;
        if (this.props.className)
            styleProps.style = this.props.style;
        this.setState({
            component: newComponent || 'div',
            props: newComponent ? ui_router_core_1.extend(props, styleProps) : styleProps,
            loaded: newComponent ? true : false
        });
    };
    UIView.prototype.registerUiCanExitHook = function (stateName) {
        typeof this.removeHook === 'function' && this.removeHook();
        var criteria = { exiting: stateName };
        var callback = this.componentInstance && typeof this.componentInstance.uiCanExit === 'function' && this.componentInstance.uiCanExit;
        if (stateName && callback) {
            var transitions = index_1.default.instance.transitionService;
            this.removeHook = transitions.onBefore(criteria, callback, {});
        }
    };
    UIView.propTypes = {
        name: react_1.PropTypes.string,
        className: react_1.PropTypes.string,
        style: react_1.PropTypes.object
    };
    UIView.childContextTypes = {
        parentUIViewAddress: react_1.PropTypes.object
    };
    UIView.contextTypes = {
        parentUIViewAddress: react_1.PropTypes.object
    };
    return UIView;
}(react_1.Component));
exports.UIView = UIView;
